File: mvp-java/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/userdetails/UserDetailsPresenter.java
package com.wasisto.githubuserfinder.android.ui.userdetails;

import com.wasisto.githubuserfinder.common.Logger;
import com.wasisto.githubuserfinder.domain.models.User;
import com.wasisto.githubuserfinder.domain.usecases.GetUserUseCase;

import java.util.concurrent.ExecutorService;

public class UserDetailsPresenter {

    private UserDetailsView view;

    private GetUserUseCase getUserUseCase;

    private Logger logger;

    private ExecutorService ioExecutorService;

    private ExecutorService mainExecutorService;

    private User user;

    public UserDetailsPresenter(UserDetailsView view, GetUserUseCase getUserUseCase, Logger logger,
                                ExecutorService ioExecutorService, ExecutorService mainExecutorService) {
        this.view = view;
        this.getUserUseCase = getUserUseCase;
        this.logger = logger;
        this.ioExecutorService = ioExecutorService;
        this.mainExecutorService = mainExecutorService;
    }

    public void load(String username) {
        mainExecutorService.execute(() -> {
            view.showLoadingIndicator();

            ioExecutorService.execute(() -> {
                try {
                    user = getUserUseCase.execute(username);

                    logger.debug(String.format("user: %s", user));

                    mainExecutorService.execute(() -> {
                        view.hideLoadingIndicator();
                        view.showUser(user);
                    });
                } catch (Exception e) {
                    logger.warn("An error occurred while getting user details", e);

                    mainExecutorService.execute(() -> {
                        view.showErrorToast();
                        view.closeActivity();
                    });
                }
            });
        });
    }

    public void onBlogClick() {
        mainExecutorService.execute(() -> {
            if (user != null && user.getBlog() != null) {
                view.openBrowser(user.getBlog());
            }
        });
    }
}

File: mvp-kotlin/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/userdetails/UserDetailsPresenter.kt
package com.wasisto.githubuserfinder.android.ui.userdetails

import com.wasisto.githubuserfinder.common.Logger
import com.wasisto.githubuserfinder.domain.models.User
import com.wasisto.githubuserfinder.domain.usecases.GetUserUseCase
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext

class UserDetailsPresenter(
    private val view: UserDetailsView,
    private val getUserUseCase: GetUserUseCase,
    private val logger: Logger,
    private val ioDispatcher: CoroutineDispatcher,
    private val mainDispatcher: CoroutineDispatcher,
    private val coroutineScope: CoroutineScope
)  {

    private lateinit var user: User

    fun load(username: String) = coroutineScope.launch(mainDispatcher) {
        view.showLoadingIndicator()

        launch(ioDispatcher) {
            try {
                user = getUserUseCase(username)

                logger.debug("user: $user")

                launch(mainDispatcher) {
                    view.hideLoadingIndicator()
                    view.showUser(user)
                }
            } catch (e: Exception) {
                logger.warn("An error occurred while getting user details", e)

                launch(mainDispatcher) {
                    view.showErrorToast()
                    view.closeActivity()
                }
            }
        }
    }

    fun onBlogClick() = coroutineScope.launch(mainDispatcher) {
        user.blog?.let {
            view.openBrowser(it)
        }
    }
}